### ISSUE TITLE
[Bug]: `alter table modify column` of type `varchar` cannot convert to type `enum`

### ISSUE THREAD

**Issue by aronchanisme** [matrixorigin/matrixone#15177]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Branch Name

main

### Commit ID

6e2ff3c8c

### Other Environment Information

```Markdown
- Hardware parameters: 8c 16g
- OS type: Debian 11
- Others:
uname -a
Linux test0 5.10.0-28-amd64 #1 SMP Debian 5.10.209-2 (2024-01-31) x86_64 GNU/Linux
```

### Actual Behavior

In MO, `alter table modify column` of type `varchar` cannot convert to type `enum` as MO throws an error indicating syntax error.

```bash
mysql> use test;
Database changed
mysql> create table t1(name varchar(25));
Query OK, 0 rows affected (0.03 sec)

mysql> 
mysql> insert into t1 values ('A'),('B'),('C');
Query OK, 3 rows affected (0.01 sec)

mysql> 
mysql> select * from t1;
+------+
| name |
+------+
| A    |
| B    |
| C    |
+------+
3 rows in set (0.01 sec)

mysql> 
mysql> alter table t1 modify column name enum('A','B'); -- Expected: not ok, mysql error: ERROR 1265 (01000): Data truncated for column 'name' at row 3
ERROR 1105 (HY000): You have an error in your SQL syntax; check the manual that corresponds to your MatrixOne server version for the right syntax to use. syntax error at line 3 column 2 near " 
)";
mysql> 
mysql> alter table t1 modify column name enum('A','B','C'); -- or alter table t1 modify column name enum('A','B','C', 'D'); -- Expected: ok
ERROR 1105 (HY000): You have an error in your SQL syntax; check the manual that corresponds to your MatrixOne server version for the right syntax to use. syntax error at line 3 column 2 near " 
)";
mysql> select * from t1;
+------+
| name |
+------+
| A    |
| B    |
| C    |
+------+
3 rows in set (0.01 sec)

```

### Expected Behavior

In MySQL 8.0, converting a column of type varchar to enum, if:
Condition-1. The list of enum contains all current values in that column, the change request will be accpected.
Condition-2. Request will be denied if condition 1 is not met.

```bash
mysql> use test;
Database changed
mysql> 
mysql> create table t1(name varchar(25));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t1 values ('A'),('B'),('C');
Query OK, 3 rows affected (0.01 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> select * from t1;
+------+
| name |
+------+
| A    |
| B    |
| C    |
+------+
3 rows in set (0.00 sec)

mysql> alter table t1 modify column name enum('A','B'); -- Expected: not ok
ERROR 1265 (01000): Data truncated for column 'name' at row 3
mysql> alter table t1 modify column name enum('A','B','C'); -- or alter table t1 modify column name enum('A','B','C', 'D'); -- Expected: ok
Query OK, 3 rows affected (0.04 sec)
Records: 3  Duplicates: 0  Warnings: 0
```

### Steps to Reproduce

```Markdown
create database if not exists test;
use test;
drop table if exists t1;
create table t1(name varchar(25));
insert into t1 values ('A'),('B'),('C');
select * from t1;
alter table t1 modify column name enum('A','B'); -- Expected: not ok, mysql error: ERROR 1265 (01000): Data truncated for column 'name' at row 3
alter table t1 modify column name enum('A','B','C'); -- or alter table t1 modify column name enum('A','B','C', 'D'); -- Expected: ok
select * from t1;
show create table t1;
desc t1;
```

### Additional information

Same priority as https://github.com/matrixorigin/matrixone/issues/15076 since from the same source.

**Comment by aronchanisme** [comment]
Fixed in main-`74464bab5`.
@Ariznawlll  Hi could you please kindly help add some bvt cases? Refer to 'steps to reproduce' for more info. Thanks.

Test result:
```bash
aronchen@momac mo % mo_ctl get_branch
2024-04-02 12:00:41.643 UTC+0800    [INFO]    Try get mo branch
2024-04-02 12:00:41.704 UTC+0800    [INFO]    Get branch succeeded, current branch: main
aronchen@momac mo % mo_ctl get_cid
2024-04-02 12:00:43.539 UTC+0800    [INFO]    Try get mo commit id
commit 74464bab51abe652c7cf77eb70ab45d66f437abb
Author: YANGGMM <www.yangzhao123@gmail.com>
Date:   Tue Apr 2 11:32:36 2024 +0800

    support create table like stmt (#15232)
    
    support create table like stmt
    
    Approved by: @iamlinjunhong, @ouyuanning, @heni02
2024-04-02 12:00:43.598 UTC+0800    [INFO]    Get commit id succeeded
aronchen@momac mo % mo_ctl connect
2024-04-02 12:00:45.243 UTC+0800    [INFO]    Checking connectivity
2024-04-02 12:00:45.292 UTC+0800    [INFO]    Ok, connecting for user ... 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 439
Server version: 8.0.30-MatrixOne-v1.1.2 MatrixOne

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> select git_version();
+---------------+
| git_version() |
+---------------+
| 74464bab5     |
+---------------+
1 row in set (0.00 sec)

mysql> create database if not exists test;
Query OK, 1 row affected (0.01 sec)

mysql> use test;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> drop table if exists t1;
Query OK, 0 rows affected (0.01 sec)

mysql> create table t1(name varchar(25));
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t1 values ('A'),('B'),('C');
Query OK, 3 rows affected (0.01 sec)

mysql> select * from t1;
+------+
| name |
+------+
| A    |
| B    |
| C    |
+------+
3 rows in set (0.01 sec)

mysql> alter table t1 modify column name enum('A','B'); -- Expected: not ok, mysql error: ERROR 1265 (01000): Data truncated for column 'name' at row 3
ERROR 20101 (HY000): internal error: convert to MySQL enum failed: item C is not in enum [A B]
mysql> alter table t1 modify column name enum('A','B','C'); -- or alter table t1 modify column name enum('A','B','C', 'D'); -- Expected: ok
Query OK, 0 rows affected (0.01 sec)

mysql> select * from t1;
+------+
| name |
+------+
| A    |
| B    |
| C    |
+------+
3 rows in set (0.01 sec)

mysql> show create table t1;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
`name` ENUM('A','B','C') DEFAULT NULL
) |
+-------+-------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> desc t1;
+-------+-------------------+------+------+---------+-------+---------+
| Field | Type              | Null | Key  | Default | Extra | Comment |
+-------+-------------------+------+------+---------+-------+---------+
| name  | ENUM('A','B','C') | YES  |      | NULL    |       |         |
+-------+-------------------+------+------+---------+-------+---------+
1 row in set (0.01 sec)
```

**Comment by aronchanisme** [comment]
Fixed in 1.1-dev(`8cd93a370`):

```bash
aronchen@momac mo % mo_ctl get_branch
2024-04-02 12:28:49.114 UTC+0800    [INFO]    Try get mo branch
2024-04-02 12:28:49.164 UTC+0800    [INFO]    Get branch succeeded, current branch: 1.1-dev
aronchen@momac mo % mo_ctl get_cid
2024-04-02 12:28:50.854 UTC+0800    [INFO]    Try get mo commit id
commit 8cd93a370a61f666fa5ffcb5efadd304f24bd7aa
Author: brown <endeavorjia@gmail.com>
Date:   Tue Apr 2 11:35:18 2024 +0800

    CI: fix 1.1 dev ci (#15281)
2024-04-02 12:28:50.915 UTC+0800    [INFO]    Get commit id succeeded
aronchen@momac mo % mo_ctl connect
2024-04-02 12:28:52.566 UTC+0800    [INFO]    Checking connectivity
2024-04-02 12:28:52.618 UTC+0800    [INFO]    Ok, connecting for user ... 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 69
Server version: 8.0.30-MatrixOne-v1.1.2 MatrixOne

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> select git_version();
+---------------+
| git_version() |
+---------------+
| 8cd93a370     |
+---------------+
1 row in set (0.00 sec)

mysql> create database if not exists test;
Query OK, 1 row affected (0.01 sec)

mysql> use test;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> drop table if exists t1;
Query OK, 0 rows affected (0.01 sec)

mysql> create table t1(name varchar(25));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t1 values ('A'),('B'),('C');
Query OK, 3 rows affected (0.00 sec)

mysql> select * from t1;
+------+
| name |
+------+
| A    |
| B    |
| C    |
+------+
3 rows in set (0.00 sec)

mysql> alter table t1 modify column name enum('A','B'); -- Expected: not ok, mysql error: ERROR 1265 (01000): Data truncated for column 'name' at row 3
ERROR 20101 (HY000): internal error: convert to MySQL enum failed: item C is not in enum [A B]
mysql> alter table t1 modify column name enum('A','B','C'); -- or alter table t1 modify column name enum('A','B','C', 'D'); -- Expected: ok
Query OK, 0 rows affected (0.01 sec)

mysql> select * from t1;
+------+
| name |
+------+
| A    |
| B    |
| C    |
+------+
3 rows in set (0.00 sec)

mysql> show create table t1;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
`name` ENUM('A','B','C') DEFAULT NULL
) |
+-------+-------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> desc t1;
+-------+-------------------+------+------+---------+-------+---------+
| Field | Type              | Null | Key  | Default | Extra | Comment |
+-------+-------------------+------+------+---------+-------+---------+
| name  | ENUM('A','B','C') | YES  |      | NULL    |       |         |
+-------+-------------------+------+------+---------+-------+---------+
1 row in set (0.01 sec)
```
